onnxscript/ir/_core.py

```diff
@@ -159,8 +172,64 @@ def display(self, *, page: bool | None = None) -> None:
             rich.print(text)
 
 
+def _check_numpy_storage_type(array: np.ndarray, dtype: _enums.DataType) -> None:
+    """Check if the numpy array dtype matches the IR data type.
+
+    For numpy unsupported dtypes, the array must be ``np.float32`` for ``BFLOAT16``,
```

Comment:
What is the goal/intention here? Eg., why not uint8 or uint16 (as in the file _type_casting.py down below), if the intention is to just have a representation of the bytes? If it is not meant to be a representation of the bytes, what is it used for?

Reply:
I updated to docstring, the thought was:

The reason for requiring ``np.float32`` for unsupported dtypes, instead of
a byte representation like uint16 or uint8, is that users will not need to provide
the btype representation for the array when they want to initialize a Tensor (when there's no numpy support for these types so it can be an annoyance),
and we will not need to cast the bytes back to float32 when users call :meth:`Tensor.numpy` to obtain the array for computation and analysis.

The correct byte representation is only needed when the tensor is serialized,
which is done in the :meth:`Tensor.tobytes` method.

Reply:
Trying to understand what usage scenarios are likely to arise.
* Post-training quantization? I guess users might have a float32 tensor that they would like to quantize. In scenarios like this, the best approach would be to encode the conversion (quantization or cast, for example) explicitly in the onnx model. With such an explicit representation of the cast/quantization, we don't need an implicit representation of such a cast internally (which is what the above seems to be supporting).
* Of course, if we have such an explicit representation, we should think about what constant-propagation would do. I would hope that (if the underlying reference implementation enables it), we could generate the bytes _when_ needed. Constant-propagation may make a decision on the profitability/usefulness of doing this.

Just thinking aloud here

Reply:
Thinking about this a bit more:

- Originally I was thinking float32 is an easier type to work with, since the ONNX helpers are all float32<->target type conversion methods. 
- Now that I have looked into the implementation of those helpers in https://github.com/onnx/onnx/pull/6124, I realize they are not performant enough or robust enough. Having a byte representation, as the onnx reference runtime does to remove the responsibility of ensuring correctness of type conversion from this tensor class will make it more robust and predictable
- So I am going to store the byte representation (uint16 for 16 bit types, uint8 for 8-bit and 4-bit types) instead. Thanks!